### PR TITLE
fix(engine-core): dynamic class with empty string results in hydration mismatch

### LINKED ISSUE #4661: Dynamic `class` with empty string results in hydration mismatch
Using `class={classes}` where `classes` is the empty string (`''`) results in a hydration mismatch after SSR+hydration.

Repro (https://github.com/salesforce/lwc/commit/13e43593052b93f2f05b5b16bfd34aa80ea6ed5d):

Result:

            Expected <p>...</p> to be <p>...</p>

We should probably check for the empty string in `style` as well.

### PR DESCRIPTION
**Pull request by cardoso**
## Details

Hopefully this is right, but if not there should be some test that fails.

Closes #4661

## Does this pull request introduce a breaking change?

-   😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

-   🔬 Yes, it does include an observable change.

No hydration mismatch on dynamic class with empty string 😄

### CODE REVIEW

**Review by nolanlawson — APPROVED** [review]

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 1]
```diff
@@ -664,7 +664,7 @@ function validateClassAttr(
 
     const elmClassName = getAttribute(elm, 'class');
 
-    if (!isUndefined(className) && String(className) !== elmClassName) {
+    if (!isUndefined(className) && String(className) !== elmClassName && className !== '') {
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:667]: This isn't totally right. If the client class name is `''` but the server class name is `'foo'` then there will be no mismatch when there should be.

Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/hydration.ts:667]: See: https://github.com/salesforce/lwc/pull/4698

**Review by nolanlawson — CHANGES_REQUESTED** [review]
Please see above

File: packages/@lwc/engine-core/src/framework/hydration.ts (hunk 1 above)
Comment (cardoso) [reply packages/@lwc/engine-core/src/framework/hydration.ts:667]: Thanks! I figured some case wasn't covered here.

**Comment by cardoso** [comment]
Thanks @nolanlawson . Should I try fixing this once you merge the new tests? Or are you taking care of it as well?

**Comment by nolanlawson** [comment]
@cardoso Feel free to go ahead! If you don't get to it, I'll get to it eventually. 🙂

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 2]
```diff
@@ -664,7 +664,12 @@ function validateClassAttr(
 
     const elmClassName = getAttribute(elm, 'class');
 
-    if (!isUndefined(className) && String(className) !== elmClassName) {
+    if (
+        !isUndefined(className) &&
+        String(className) !== elmClassName &&
+        // No mismatch if SSR className is empty and the client-side class is null
+        !(className === '' && isNull(elmClassName))
+    ) {
```
Comment (cardoso) [inline packages/@lwc/engine-core/src/framework/hydration.ts:672]: This if is starting to get hairy, but the auto-formatting gave me the idea of adding this comment. I think this is OK.

**Review by cardoso — COMMENTED** [review]
Fixed the logic, tests are passing @nolanlawson .

File: packages/@lwc/engine-core/src/framework/hydration.ts (hunk 2 above)
Comment (cardoso) [reply packages/@lwc/engine-core/src/framework/hydration.ts:672]: Could also be `(className !== '' || !isNull(elmClassName))` but I found this way more readable and easy to explain via the comment.

**Review by nolanlawson — APPROVED** [review]
Looks great! Thank you. 🙇

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 3]
```diff
@@ -664,7 +664,12 @@ function validateClassAttr(
 
     const elmClassName = getAttribute(elm, 'class');
 
-    if (!isUndefined(className) && String(className) !== elmClassName) {
+    if (
+        !isUndefined(className) &&
+        String(className) !== elmClassName &&
+        // No mismatch if SSR className is empty and the client-side class is null
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:670]:
```suggestion
        // No mismatch if SSR `class` attribute is missing and CSR `class` is the empty string
```

**Pull request merged**